### DISCUSSION TITLE
React 18 support

### DISCUSSION THREAD

**Discussion by wardpeet** [gatsbyjs/gatsby#31943] [Umbrella Discussions] [answered]
The React team has announced the [Stable version of React 18](https://reactjs.org/blog/2022/03/29/react-v18.html). We're part of the working groups and are adopting new features, one by one.

## How to use

Install `react@latest` and `react-dom@latest`. We will automatically switch to the new CreateRoot API.

## Your feedback

We're very much looking for feedback from you using React 18 and its new features. We'll expand this section with more asks and details in the future, but for now, our specific asks would be:

- First and foremost, does your site fundamentally work?
- Any libraries/plugins that do not work with this approach?

## Notes

You're now able to use new functionality like `startTransition`, `stateBatching`, ...

Check out our [React 18 e-commerce demo](https://react18ecommercedemo.gatsbyjs.io/), the code lives at https://github.com/gatsbyjs/react-18-ecommerce-demo

**Comment by mwskwong** [comment]
It appears that part of the application is still using `ReactDOM.render`. I'm getting the following error constantly:
![image](https://user-images.githubusercontent.com/32725014/122711607-04f42000-d295-11eb-87d2-ea5cba7ab3df.png)

<details>
  <summary>Gatsby Info</summary>

```
  System:
    OS: Windows 10 10.0.18363
    CPU: (12) x64 Intel(R) Core(TM) i7-8700 CPU @ 3.20GHz
  Binaries:
    Node: 14.17.0 - C:\Program Files\nodejs\node.EXE
    Yarn: 1.22.5 - C:\Program Files (x86)\Yarn\bin\yarn.CMD
    npm: 7.14.0 - C:\Program Files\nodejs\npm.CMD
  Languages:
    Python: 3.9.5
  Browsers:
    Chrome: 91.0.4472.114
    Edge: Spartan (44.18362.449.0)
  npmPackages:
    gatsby: ^3.7.2 => 3.7.2 
    gatsby-plugin-emotion: ^6.7.1 => 6.7.1 
    gatsby-plugin-image: ^1.7.1 => 1.7.1 
    gatsby-plugin-manifest: ^3.7.1 => 3.7.1 
    gatsby-plugin-offline: ^4.7.1 => 4.7.1 
    gatsby-plugin-react-helmet: ^4.7.1 => 4.7.1 
    gatsby-plugin-robots-txt: ^1.6.2 => 1.6.2 
    gatsby-plugin-root-import: ^2.0.6 => 2.0.6 
    gatsby-plugin-sharp: ^3.7.1 => 3.7.1 
    gatsby-plugin-sitemap: ^4.3.1 => 4.3.1 
    gatsby-plugin-webpack-bundle-analyser-v2: ^1.1.22 => 1.1.21 
    gatsby-source-filesystem: ^3.7.1 => 3.7.1 
    gatsby-transformer-sharp: ^3.7.1 => 3.7.1 
```

</details>
<details>
  <summary>Gatsby Flags</summary>

```js
{
    FAST_DEV: true,
    DEV_SSR: true,
    PRESERVE_FILE_DOWNLOAD_CACHE: true,
    PRESERVE_WEBPACK_CACHE: true
}
```

</details>

**Edit:** I believe it is caused by `gatsby-plugin-image`. As soon as remove/comment out all instances of `gatsby-plugin-image`, the warning/error is not prompted again.

**Reply by LekoArts** [reply]
What happens when you use `DEV_SSR: false`?

**Reply by mwskwong** [reply]
> What happens when you use `DEV_SSR: false`?

Same. In fact, the same warning is prompted even when no flags are used.

**Reply by wardpeet** [reply]
I think it's gatsby-plugin-image, are you using it?

**Reply by mwskwong** [reply]
@wardpeet As mentioned in the edit of my OP, that's exactly the case.

**Reply by mwskwong** [reply]
Is there any road map on when gatsby-plugin-image supports React 18 and the new `createRoot` function?

**Reply by mwskwong** [reply]
I believe you guys should have already aware of this, the latest stable version as of today (i.e. `1.9.0`) of `gatsby-plugin-image` is still using the old `render()` function.

**Reply by wardpeet** [reply]
I'm aware, I'll fix it asap, It's more a warning than a hard fail right?

**Reply by mwskwong** [reply]
Yeah, it doesn't break the build at least.

**Reply by mwskwong** [reply]
@wardpeet I'm running out of ideas. I noticed some (yes, not all) of the components that make use of `gatsby-plugin-image` in my project will cause the DEV build to crash and disconnect the socket. While for some reason, I'm unable to reproduce it with a minimal setup. This can be workarounded by downgrading `gatsby-plugin-sharp` to 3.8.

My repo: https://github.com/matthewkwong2/resume-gatsby/tree/next
The component that works: https://github.com/matthewkwong2/resume-gatsby/blob/next/src/sections/index/Work/CompanyLogo.js
One of the components that crashes the DEV build: https://github.com/matthewkwong2/resume-gatsby/blob/next/src/sections/index/Project/ProjectCard.js

**Reply by wardpeet** [reply]
Is there anything in particular that you're doing? it's not crashing for me

**Reply by mwskwong** [reply]
I'm not sure which dependency might have caused it, but it seems that it has been working fine after a few days since I post that comment. I'm updating my dependency daily though.

**Reply by mwskwong** [reply]
I saw there is a PR https://github.com/gatsbyjs/gatsby/pull/32378 merged 3 days ago which is supposed to fix this issue. Yet, there is no new `next` release for this. May I know is there a schedule for the next `next` and/or public release?

**Reply by stevepepple** [reply]
I upgraded gatsby-plugin-image and gatsby-plugin-sharp, but still see the warning. Are there any other plugins I should watch out for?

**Reply by rauscht** [reply]
dasda @moseso

**Comment by mwskwong** [comment]
Another question will be, with React 18, is it possible to pre-render/SSR the entire page so that the page will have all the barebone HTML elements by default. Then, when it comes to re-hydration, it "Selective Hydrate", prioritizing those sections (wrapped in `Suspense`) which are either in view or users perform interaction on them?

Currently, when using `Suspense`, Gatsby (and properly React's fault) will display the fallback by default, while in many cases, the barebone, pre-rendered version of the components are the best "fallback".

**Reply by wardpeet** [reply]
We'll be having support for suspense in SSR in version 3.9.

**Reply by mwskwong** [reply]
> We'll be having support for suspense in SSR in version 3.9.

That's really good news. May I know is that available in 3.9 beta now?

**Reply by wardpeet** [reply]
not yet, I'll get a PR up and running today or tomorrow :)

**Comment by mwskwong** [comment]
So I notice there is a release on npm named `3.9.0-alpha-ssr.15`. If I'm not mistaken, this is the version that includes Suspense SSR's support with React 18. However, there are further releases of `3.6.0-alpha-ssr.xxx`. Despite the version number appears to be an outdated version, it is newer than `3.9.0-alpha-ssr.15`.

Does anyone know what is going on? I'm confused.

**Reply by wardpeet** [reply]
It's something completely different, suspense support is going to land in 3.9. You can already try it if you install `gatsby@next`

**Reply by mwskwong** [reply]
> It's something completely different, suspense support is going to land in 3.9. You can already try it if you install `gatsby@next`

Hmm... I did install `gatsby@next` (which for some reason is pointing to `3.10 next`) and when I use `Suspense` and `lazy`, the entire component is not SSR. Am I supposed to enable some sort of settings (e.g. flags) to make it work?

**Reply by wardpeet** [reply]
Thanks, I've published a new @next version and now it's part of it.

**Reply by mwskwong** [reply]
Tested and it works great. I no longer need 3rd party library to achieve component-based code-splitting.

**Comment by mwskwong** [comment]
So after using `gatsby-plugin-image@1.11.1-next`, most of the `ReactDOM.render` warnings are gone. Yet, there is still one that remains, and it seems to originate from `app.js` `runRender` function. No flag used BTW. Would it be possible to relate to `DEV_SSR`?
![image](https://user-images.githubusercontent.com/32725014/126246351-711aadcb-a7f6-403f-bdf2-fd4b5b42ac49.png)


**Edit:** It does seem to relate to `DEV_SSR`. Set it to `true` and the above warning can no longer be seen (I thought it is enabled by default already?). Obviously, setting `FAST_DEV` to `true` will do the same thing.

**Comment by ardiewen** [comment]
Just wanted to chime in with another potential issue:

Using `gatsby-plugin-image` with react 18 seems to create some conflicts:

1. Still getting this message on latest libraries as of today (see below for exact environment):
```
Warning: ReactDOM.render is no longer supported in React 18. Use createRoot instead. Until you switch to the new API, your app will behave as if it's running React 17. Learn more: https://reactjs.org/link/switch-to-createroot
```

2. Unmounting art directed images seems to cause problems in the dev environment:
![image](https://user-images.githubusercontent.com/39966887/141714473-4af400b5-a638-4817-8f46-8db3ef632f16.png)

Minimal repro here:
https://github.com/ardiewen/gatsby-image-react18-repro

Page `/` is art directed and works as expected.

Page `/carousel` is not art directed showing compatibility between a carousel library and non-art directed images.

Navigating using gatsby-link from `/` to `/carousel` will cause a runtime error described above.

Page `/artdirected` combines the carousel library with art directed images. When the carousel is dragged, it causes the same issue as navigating using gatsby-link. 

Downgrading to React 17 causes the runtime errors to go away on `, but the images still "flicker" when swiped through on the art directed carousel (the non art directed carousel doesn't flicker).

Gatsby Info dump:
```
  System:
    OS: Linux 5.10 Ubuntu 20.04.3 LTS (Focal Fossa)
    CPU: (16) x64 AMD Ryzen 7 5800X 8-Core Processor
    Shell: 5.8 - /usr/bin/zsh
  Binaries:
    Node: 14.17.5 - ~/.nvm/versions/node/v14.17.5/bin/node
    Yarn: 1.22.15 - /usr/bin/yarn
    npm: 6.14.14 - ~/.nvm/versions/node/v14.17.5/bin/npm
  npmPackages:
    gatsby: ^4.1.4 => 4.1.4
    gatsby-plugin-image: ^2.1.2 => 2.1.2
    gatsby-plugin-sharp: ^4.1.3 => 4.1.3
    gatsby-plugin-styled-components: ^5.1.0 => 5.1.0
    gatsby-source-filesystem: ^4.1.2 => 4.1.2
    gatsby-transformer-sharp: ^4.1.0 => 4.1.0
```

**Comment by phil-lgr** [comment]
I was on 

```
    "react": "18.0.0-alpha-55d75005b-20211011",
    "react-dom": "18.0.0-alpha-55d75005b-20211011",
```

with Gatsby version:

```
    "gatsby": "^3.14.1",
```

and it is working fine, except some minor hydration errors

I just tried with:

```
    "react": "^18.0.0-beta-149b420f6-20211119",
    "react-dom": "^18.0.0-beta-149b420f6-20211119",
```

And got the same message as ardiewen (https://github.com/gatsbyjs/gatsby/discussions/31943#discussioncomment-1642200)

Are there any news for React beta v18 compatibility with Gatsby v3 or v4?

Thanks a lot, 

Phil

**Reply by phil-lgr** [reply]
I also noticed that the demo here : https://www.gatsbyjs.com/blog/how-to-try-react-18-in-gatsby/ has images missing

**Comment by mosesoak** [comment]
Hi @wardpeet and Gatsby team, we also started a minimal project and got the same error,
`react_devtools_backend.js:4045 Warning: ReactDOM.render is no longer supported in React 18. Use createRoot instead. Until you switch to the new API, your app will behave as if it's running React 17.`
![image](https://user-images.githubusercontent.com/24345/150003510-454d43b4-816f-4928-95fd-a1f2a7a055c8.png)

This happens with no flags, and using the latest releases as of today, `"gatsby": "^4.5.3", "react": "^18.0.0-rc.0-next-fe905f152-20220107"`

Our team is very interested in helping you get this off the ground, as our sites tend to be CPU-heavy despite existing optimization options, so we're hungry for the new Transition API. We do use Netlify CMS and rely heavily on `gatsby-plugin-image` in our work.

Any progress report on this front? And thanks for the effort to break this ground!

**Comment by EarningsCall** [comment]
We are currently currently using React 18 for SSR with `Suspense` and React's `Lazy` support to lazy load components.

We have tried upgrading multiple times to Gatsby 4.X but have run into issues every time.  It seems that Gatsby 4 doesn't support React 18, and we are therefore stuck on Gatsby 3.14.

Is there any plan to continue React 18 support?

**Comment by tyhopp** [comment, marked as answer]
Hi all, as of the [recent release of gatsby 4.11](https://www.gatsbyjs.com/docs/reference/release-notes/v4.11/#react-18) we have 100% compatibility with React 18 again. I encourage you to test it out and share feedback here with your experience.

**Reply by mwskwong** [reply]
And I did. Everything works fine except `gatsby-plugin-offline`, which I mentioned in https://github.com/gatsbyjs/gatsby/issues/35193

**Reply by tyhopp** [reply]
> And I did. Everything works fine except `gatsby-plugin-offline`, which I mentioned in #35193

Thanks for resurfacing this

**Reply by josehernandezv** [reply]
I'm still getting **ReactDOM.render is no longer supported in React 18** with Gatsby 4.12.1

**Comment by davidpaulsson** [comment]
After updating to 4.11 and bumping react to 18, I'm getting these `Hydration failed because the initial UI does not match what was rendered on the server.` and `There was an error while hydrating. Because the error happened outside of a Suspense boundary, the entire root will switch to client rendering.` errors

**Reply by mwskwong** [reply]
Using `gatsby-plugin-offline`? If yes, then it may be the same case as https://github.com/gatsbyjs/gatsby/issues/35193

**Reply by davidpaulsson** [reply]
@mwskwong I'm not :(

**Reply by mwskwong** [reply]
For me, I was using [Intl](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl) without specifying the locale. And that causes different values to be rendered. Not sure if you are using any client-side-only API that perhaps you are not aware of?

**Reply by LekoArts** [reply]
Can you share a minimal reproduction?
https://www.gatsbyjs.com/contributing/how-to-make-a-reproducible-test-case/

**Reply by badaczewski** [reply]
Not sure if this is related but after upgrading to React 18 in my Gatsby 4.1.3 project I'm getting these errors:

**Module not found: Error: Package path ./compat/client is not exported from package D:\Gatsby\frontend\node_modules\preact (see exports field in D:\XXX\XXX\node_modules\preact\package.json)**

I've updated all the preact packages to the latest, but still getting these errors during the build/develop process, specfically during gatsby-plugin-image processing.

**Reply by duncanf005** [reply]
**Module not found: Error: Package path ./compat/client is not exported from package D:\Gatsby\frontend\node_modules\preact (see exports field in D:\XXX\XXX\node_modules\preact\package.json)**

Having the same issue here :(

**Reply by darmentrout** [reply]
I'm getting the same error about hydration ([#418](https://reactjs.org/docs/error-decoder.html/?invariant=418)). 
"gatsby": "^4.13.1",
"react": "^18.1.0",

**Reply by mwskwong** [reply]
OK, I retested a bit and it seems to be related to the version of the browsers.

On Chrome 97, no issues
On Chrome 98, the hydration error can be observed, although doesn't really affect the functionality in any way.
On Chrome 101, no issues.

Note my `borwserslist` config is
```json
[
  "> 0.5%",
  "last 2 versions",
  "Firefox ESR",
  "not dead",
  "not IE 11"
]
```

So in theory, Chrome 97 and 98 are not supported by my website.

**Reply by ashhitch** [reply]
We are getting this issue with React 18...  without the offline plugin.
Did anyone find a solution?

**Reply by hendra-go** [reply]
This bug occured because [react 18 breaking change: **Stricter hydration errors**](https://reactjs.org/blog/2022/03/08/react-18-upgrade-guide.html#other-breaking-changes)

For me, temporary fix is by forcefully render using `createRoot` instead of `hydrateRoot` (info: [React 18 Updates to Client Rendering APIs](https://reactjs.org/blog/2022/03/08/react-18-upgrade-guide.html#updates-to-client-rendering-apis))

If your `gatsby-browser.js` use CommonJS
```
const reactDomClient = require(`react-dom/client`)

exports.replaceHydrateFunction = () => {
  return (element, container) => {
    reactDomClient.createRoot(container).render(element)
  };
};
```

Or if your `gatsby-browser.js` use ES6
```
import { createRoot } from `react-dom/client`

export const replaceHydrateFunction = () => {
  return (element, container) => {
    createRoot(container).render(element)
  }
}
```
This is just a temporary fix until DEV_SSR return on the next gatsby version.

**Reply by mwskwong** [reply]
You are essentially disabling SSR entirely in this "temporary fix". What's the point of that?

**Reply by hendra-go** [reply]
> You are essentially disabling SSR entirely in this "temporary fix". What's the point of that?

Well, from what i understand, by using `createRoot` we will not be able to use **SSR in development feature**. DEV_SSR currently disabled anyway. DEV_SSR was disabled shortly before v4 release due to `getServerData` not being handled correctly. see - [#33504 comment](https://github.com/gatsbyjs/gatsby/pull/33054#discussion_r707189586)

**Reply by mwskwong** [reply]
The thing is, the DEV build is already using CSR in Gatsby v4 from the very beginning. What you are doing here is making the PROD build also using CSR since you didn't use any conditions to check the built environment.

Not to mention that the "error" people mentioned here will only appear in PROD build, which as the message stated, React uses CSR as a fallback anyway. 

So you didn't fix anything here. In fact, you just force everything into the "fallback mode".

**Reply by hendra-go** [reply]
> So you didn't fix anything here. In fact, you just force everything into the "fallback mode".

You are right about not fixing anything.

Ideally we should hunt down Hydration mismatches bugs and fix them, but the bugs location hard to find without DEV_SSR, hopefully DEV_SSR will return soon see #35685

At first, my project is run smoothly without hydration error on production build (using latest gatsby and also gatsby-plugin-offline)

Hydration error come up after i use `gatsby-plugin-minify` which modified the generated html `onPostBuild`

I fixed the hydration error on my production build by using `replaceHydrateFunction` above.

I think is fair not to call it a "fix", a "work around" should be more appropriate.

So i just want to share the "work around" to remove hydration error on production build which similiar to my use case (the generated html modified `onPostBuild`)

**Reply by mwskwong** [reply]
> I fixed the hydration error on my production build by using replaceHydrateFunction above.

You still don't get the point. You are forcing your entire application to use CSR in production with your "fix"/"workaround". Of course, you won't see any hydration errors because you are no longer using hydration now.

Will it work? Yes. But logically, will you force the ENTIRE APP to use CSR instead of allowing it to fall back to CSR in PART OF THE APPLICATION? No.

**Reply by GilliganL** [reply]
We are currently experiencing this issue on Gatsby v4.17.0, React 18.2.0. We are using gatsby-plugin-offline.

**Reply by tadeas22** [reply]
We are currently experiencing same issue on Gatsby v4.19.1, React 18.2.0, We are NOT using gatsby-plugin-offline.

**Reply by nnmax** [reply]
We are currently experiencing same issue on Gatsby v4.21.1, React 18.2.0, We are NOT using gatsby-plugin-offline.

**Reply by bared-dave** [reply]
We have recently upgraded to v4 from v2, and even paring the site right back to its most basic, I still for the life of me can't figure out why this error is occurring. It seems to be even introducing the most basic hook causes the error. As others have observed, it can sometimes be browser dependent. Rarely see it in Chrome, see it every time in Firefox.

**Reply by natapg** [reply]
same issue here with gatsby 4.23.1 and react 18.2.0, we are NOT using gatsby-plugin-offline.

**Reply by Ozerich** [reply]
same issue here with gatsby 4.24 and react 18.2.0, we are NOT using gatsby-plugin-offline.

**Reply by jalal246** [reply]
I had the same bug and it affected the site badly because eventually, it re-render the entire page. Not one component or section, the entire landing page loaded then it flashes and appears again.
I couldn't investigate the issue. When I build locally it's fine, when using Gatsby cloud it appears. Unfortunately, Gatsby doesn't provide a formal way to debug the production bundle. So I added `process.env.NODE_ENV = "development"` in `gatsby-config` file. That allowed me to debug React error in the development mode (Gatsby bundled still minified) and it turns out there is a mismatch in the text. I have a date that is rendered differently causing text mismatch and eventually killing the first load.
Adding `timeZone` to the date solves the problem while apparently pushing another bug, that again doesn't appear when running  `develop` or `build` locally but only when deploying with a cloud server.

**Reply by xandor-io** [reply]
I was dealing with the same issue after updating my app from gatsby v2 to v4. After reading some documentation I came across the `DEV_SSR: true` flag in gatsby-config.js. The feature enables SSR in development so certain bugs that previously only show up in builds (like trying to access the window object) will now be seen during development which will help speed up fixing them. [See Discussion Here](https://github.com/gatsbyjs/gatsby/discussions/28138)

```
flags: {
  DEV_SSR: true
},
```

This allowed me to test further in my dev environment. My issue had to do with `alt` tags not matching what was rendered on the server. After further investigation it was due to using `gatsby-image`. I switch to `gatsby-plugin-image` and this seem to resolve this issue. I follow the migration docs [here](https://www.gatsbyjs.com/docs/reference/release-notes/image-migration-guide/) using this command. `npx gatsby-codemods gatsby-plugin-image `.

I hope this helps, if anyone else is running into the same issue.

**Reply by tobi-oye** [reply]
Hi i also still have this issue even with 
```
flags: {
  DEV_SSR: true
}, 
```

was anyone able to resolve this ??

**Comment by SalazarJosh** [comment]
Sorry to dig up this old post, but I'm not finding anything around React 19 support. Are there plans to continue to support the latest major version of React (v19.1 currently)?
